ceremonies/meetup-validation/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+		let (
+			participants_grouped_by_outgoing_attestations,
+			participants_grouped_by_incoming_attestations,
+		) = group_participants_by_num_incoming_and_outgoing_attestations(
+			&participants_to_process,
+			&relevant_attestations,
+		);
+
+		let min_num_outgoing_attestations = participants_grouped_by_outgoing_attestations[0].0;
+		let min_num_incoming_attestations = participants_grouped_by_incoming_attestations[0].0;
+
+		if min_num_incoming_attestations < min_num_outgoing_attestations {
+			if min_num_incoming_attestations < threshold_fn(participants_to_process.len()) {
+				let ps = participants_grouped_by_incoming_attestations[0].1.clone();
+				ps.clone().into_iter().for_each(|p| {
+					excluded_participants.push((p, ExclusionReason::TooFewIncomingAttestations))
+				});
+
+				// remove the participants from the included participants and the attestation vectors
+				participants_to_process.retain(|k| !ps.contains(k));
+				relevant_attestations =
+					filter_attestations(&participants_to_process, &relevant_attestations);
+				continue
+			}
+		} else {
+			if min_num_outgoing_attestations < threshold_fn(participants_to_process.len()) {
+				let ps = participants_grouped_by_outgoing_attestations[0].1.clone();
+				ps.clone().into_iter().for_each(|p| {
+					excluded_participants.push((p, ExclusionReason::TooFewOutgoingAttestations))
+				});
+
+				// remove the participants from the included participants and the attestation vectors
+				participants_to_process.retain(|k| !ps.contains(k));
+				relevant_attestations =
+					filter_attestations(&participants_to_process, &relevant_attestations);
+				continue
+			}
+		}
+
+		// if all participants are above the threshold and therefore no participants were removed, we exit the loop
+		break
+	}
+	excluded_participants
+}
+
+fn find_majority_vote(
+	participants: &Vec<usize>,
+	participant_votes: &Vec<u32>,
+) -> Result<(u32, u32), MeetupValidationError> {
+	let mut n_vote_candidates: Vec<(u32, u32)> = vec![];
+	for i in participants {
+		let this_vote = participant_votes[*i];
+		match n_vote_candidates.iter().position(|&(n, _c)| n == this_vote) {
+			Some(idx) => n_vote_candidates[idx].1 += 1, //safe; <= number of candidates
+			_ => n_vote_candidates.insert(0, (this_vote, 1)),
+		};
+	}
+
+	if n_vote_candidates.is_empty() {
+		return Err(MeetupValidationError::BallotEmpty)
+	}
+	// sort by descending vote count
+	n_vote_candidates.sort_by(|a, b| b.1.cmp(&a.1));
+	if n_vote_candidates[0].1 < 3 {
+		//safe; n_vote_candidate not empty checked aboveÏ
+		return Err(MeetupValidationError::NoDependableVote)
+	}
+	let (n_confirmed, vote_count) = n_vote_candidates[0];
+	Ok((n_confirmed, vote_count))
+}
+
+fn filter_attestations(
+	participants: &Vec<usize>,
+	participant_attestations: &Vec<Vec<usize>>,
+) -> Vec<Vec<usize>> {
+	// filter out participants from the attestation vectors that are not in the participants vector anymore.
+	participant_attestations
+		.clone()
```

Comment:
It is not good practice to pass references to a function when you need to call clone inside the functions. Better pass an owned value, this potentially removes the amount because the value might not be needed later on the outside. You can also be more ergonomic here:

```
fn filter_attestations(
	participants: &Vec<usize>,
	participant_attestations: Vec<Vec<usize>>,
) -> Vec<Vec<usize>> {
	// filter out participants from the attestation vectors that are not in the participants vector anymore.
	participant_attestations
		.into_iter()
		.map(|a| a.into_iter().filter(|j| participants.contains(j)).collect())
		.collect()
}
```

Reply:
Edit: Even more concise and efficient:

```
	participant_attestations
		.into_iter()
		.map(|mut a| {
			a.retain(|j| participants.contains(j));
			a
		})
		.collect()
```